### PR TITLE
feature: add version.h file and --version args, and print version and commit in log

### LINKED ISSUE #1815: Update version number on release
### Background

Currently ABACUS writes version number in source code. This requires additional effort when releasing a new version.

### Describe the solution you'd like

Maybe we should find a way to automatically configure the version info.

### Additional Context

_No response_

### LINKED ISSUE #2430: We need the version with binding commit or date
### Background

At present, ABACUS is in the stage of rapid development, and many calculations and tests are carried out using the code of the develop branch. When we get a result, it is difficult to determine the actual codes it runs. 

It is best to output a version number during the calculation, which is bound to the github's commit.

### Describe the solution you'd like

One approach is to add a version file that stores the ABACUS version number and the latest commit number. After each PR is received, github automatically changes the commit number in the version file to the latest and pushes the PR. But this approach will double the number of PR. 

Alternatively, we can bind the date instead of commit. After the first PR is received every day, github can modify the date information in the version file, and push to develop repository.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by pxlxingliang**

### CODE REVIEW

**Review by dyzheng — CHANGES_REQUESTED** [review]
1. This PR is a feature，please add Tag.
2. please add COMMIT_INFO description in documentation.

File: CMakeLists.txt [hunk 1]
```diff
@@ -33,6 +33,31 @@ option(DEBUG_INFO "Print message for developers to debug." OFF)
...
+
+#read version from VERSION file
+file(STRINGS "VERSION" VERSION)
+string(STRIP ${VERSION} VERSION)
+add_definitions("-DVERSION=\"${VERSION}\"")
+
+if(COMMIT_INFO)
+  # Get commit hash
+  execute_process(
+    COMMAND git log -1 --pretty=format:%h
```
Comment (caic99) [inline CMakeLists.txt]: Use `find_program` to find git and bash first is better

Comment (caic99) [inline CMakeLists.txt]: What if the user downloads the code from release page but not from the git repo? In this case it is unable to tell the version from git.

File: VERSION [hunk 2]
```diff
@@ -0,0 +1 @@
+v3.2.2
```
Comment (caic99) [inline VERSION]: How about setting version info in a `version.h`?
```suggestion
#ifndef VERSION
#define VERSION "v3.2.2"
#endif
```

**Comment by pxlxingliang** [comment]
> 1. This PR is a feature，please add Tag.
> 2. please add COMMIT_INFO description in documentation.

1. Done
2. Sorry, I don't know which document to add to？

File: VERSION (hunk 2 above)
Comment (pxlxingliang) [reply VERSION]: Currently, it is passed using macros. What are the benefits of placing it in the version. h file?

Comment (caic99) [reply VERSION]: > Currently, it is passed using macros. What are the benefits of placing it in the version. h file?

Just by common practice. It's OK to keep your original implementation.

**Review by dyzheng — APPROVED** [review]

File: source/main.cpp [hunk 3]
```diff
@@ -5,9 +5,21 @@
...
+    {
+#ifdef VERSION
+        const char* version = VERSION;
+#else
+        const char* version = "unknown";
+#endif
+        std::cout << "ABACUS version " << version << std::endl;
+        std::exit(0);
+    }
+
```
Comment (hongriTianqi) [inline source/main.cpp]: @pxlxingliang Please add a unit test for this new function, thanks!

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

File: VERSION (hunk 2 above)
Comment (pxlxingliang) [reply VERSION]: Ok, I accept your suggestion.

**Review by hongriTianqi — APPROVED** [review]

**Pull request merged**